### PR TITLE
Expanded Self-Query Retriever 👀 and Self-querying with MyScale 🤩

### PR DESCRIPTION
**Pull request by mpskex**
# Expanded Self-Query Retriever 👀 and Self-querying with [MyScale](https://docs.myscale.com/en/) 🤩

LangChain is built with a great interface for self-query retrievers and many super nice contribution has been made during the last few weeks. 🌟 Self-query retrievers are really cool and we believe it will be the future of vector databases! 😎 So we decided to expand self-query retriever with more functions and data types enabled and bring MyScale to the big family of self-query enabled vector databases! 🤗

[MyScale](https://docs.myscale.com/en/) can make use of [various data types and functions for filters](https://blog.myscale.com/2023/06/06/why-integrated-database-solution-can-boost-your-llm-apps/#filter-on-anything-without-constraints). It will boost up your LLM app no matter if you are scaling up your data or expand your system to broader application. 🚀🚀🚀 And we believe it will be a common feature to every vector database in the future!

In the notebook we'll demo the `SelfQueryRetriever` wrapped around a MyScale vector store with some extra piece we contributed to LangChain. In short, it can be concluded into 4 points:

1. Add `contain` comparator 🪣 to match list of any if there is more than one element matched
2. Add `timestamp` data type 🕰️ for date time match (ISO-format, or YYYY-MM-DD)
3. Add `like` comparator 🔎 for string pattern search
4. Add arbitrary function 💣 capability. 

Let's boost LLM apps with vector databases! 🎉🎉🎉

#### Before submitting

- [x] unit-test for `LIKE` and `CONTAIN` 
- [x] Notebook for MyScale Self-query Retriever

#### Who can review?

Tag maintainers/contributors who might be interested:

- @hwchase17 
- @dev2049

### CODE REVIEW

File: langchain/vectorstores/mongodb_atlas.py [hunk 1]
```diff
@@ -21,8 +21,6 @@
 if TYPE_CHECKING:
     from pymongo.collection import Collection
 
-MongoDBDocumentType = TypeVar("MongoDBDocumentType", bound=Dict[str, Any])
```
Comment (hwchase17) [inline langchain/vectorstores/mongodb_atlas.py:24]: why this change?

Comment (mpskex) [reply langchain/vectorstores/mongodb_atlas.py:24]: I got lint error on my local poetry said unbounded type var. I will try to revert back and see if CI passes.
